caiman/source_extraction/cnmf/lazy_arrays.py

```diff
@@ -0,0 +1,439 @@
+        """
+        np.dtype: numpy dtype
+        """
+        # compute the slice at the first index and return the dtype of that
+        # TODO: there might be a nicer way to do this
+        return self[0].dtype
+
+    @property
+    @abstractmethod
+    def shape(self) -> Tuple[int, int, int]:
+        """
+        Tuple[int]: (n_frames, dims_x, dims_y)
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def n_frames(self) -> int:
+        """
+        int: number of frames
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def min(self) -> float:
+        """
+        float: min value of the array if it were fully computed
+        """
+        pass
+
+    @property
+    @abstractmethod
+    def max(self) -> float:
+        """
+        float: max value of the array if it were fully computed
+        """
+        pass
+
+    @property
+    def ndim(self) -> int:
+        """
+        int: Number of dimensions
+        """
+        return len(self.shape)
+
+    @property
+    def nbytes(self) -> int:
+        """
+        int: number of bytes required for the array if it were fully computed
+        """
+        return np.prod(self.shape + (np.dtype(self.dtype).itemsize,))
+
+    @property
+    def nbytes_gb(self) -> float:
+        """
+        float: number of gigabytes required for the array if it were fully computed
+        """
+        return self.nbytes / 1e9
+
+    @abstractmethod
+    def _compute_at_indices(self, indices: Union[int, slice]) -> np.ndarray:
+        """
+        Lazy computation logic goes here. Computes the array at the desired indices.
+
+        Args:
+            indices (Union[int, slice]): the user's desired slice, i.e. slice object or int passed from `__getitem__()`
+
+        Returns:
+            np.ndarray: slice of the array at the desired indices
+
+        """
+        pass
+
+    def as_numpy(self):
+        """
+        Converts to a standard numpy array in RAM.
+
+        NOT RECOMMENDED, THIS COULD BE EXTREMELY LARGE. Check ``nbytes_gb`` first!
+
+        Returns:
+            np.ndarray: full dense array in RAM
+        """
+
+        warn(
+            f"\nYou are trying to create a numpy.ndarray from a LazyArray, "
+            f"this is not recommended and could take a while.\n\n"
+            f"Estimated size of final numpy array: "
+            f"{self.nbytes_gb:.2f} GB"
+        )
+        dense = np.zeros(shape=self.shape, dtype=self.dtype)
+
+        for i in range(self.n_frames):
+            dense[i] = self[i]
+
+        return dense
+
+    def save_hdf5(self, filename: Union[str, Path], dataset_name: str = "data"):
+        """
+        Save the full dense array as an hdf5 file.
+
+        **Note: This could result in a very large file, use ``nbytes_gb`` to check the
+        size of the full dense array**
+
+        Args:
+            filename:
+            dataset_name:
+
+        Returns:
+
+        """
+        pass
```

Comment:
presumably this is more WIP

Reply:
yes I am not sure how useful this really would be since it can be hundreds of gigabytes, thoughts @EricThomson  ?

Reply:
We can probably skip as_numpy() but save_hdf5 might be useful (provided we make sure it's also not a ram gobbler), particularly for smaller datasets.

Reply:
Actually, maybe zarr or n5 would be a better choice

Reply:
Yes! `zarr` would be the natural choice, though we would need to add that as a dependency?

Reply:
I ordinarily hate to have dynamic imports; z5py is the sole example of this in our codebase (to my knowledge) because it's not available in pip (for the few cases we care about pip). I don't mind our code generally depending on it and saying "you're missing a few features" to the people with pip-based installs (they're already off the beaten path).

Reply:
`zarr` is on both conda-forge and pypi:
https://anaconda.org/conda-forge/zarr
https://pypi.org/project/zarr/

I wonder if it's cross-compatible with `z5py`

Reply:
I like all these options. :guitar: